File: orion-viewer/src/main/java/universe/constellation/orion/viewer/OrionBookmarkActivity.java
/*
 * Orion Viewer - pdf, djvu, xps and cbz file viewer for android devices
 *
 * Copyright (C) 2011-2013  Michael Bogdanov & Co
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package universe.constellation.orion.viewer;

import android.app.Activity;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.util.SparseBooleanArray;
import android.view.*;
import android.widget.*;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;
import org.xmlpull.v1.XmlPullParserFactory;
import universe.constellation.orion.viewer.bookmarks.*;

import java.io.*;
import java.util.*;

import static universe.constellation.orion.viewer.LoggerKt.log;

/**
 * User: mike
 * Date: 25.02.12
 * Time: 16:20
 */
public class OrionBookmarkActivity extends OrionBaseActivity {

    public static final String OPEN_PAGE = "open_page";

    public static final String OPENED_FILE = "opened_file";

    public static final String BOOK_ID = "book_id";

    public static final String NAMESPACE = "";

    public static final int IMPORT_CURRRENT = 1;

    public static final int IMPORT_ALL = 2;

    private long bookId;

    public void onCreate(Bundle savedInstanceState) {
        super.onOrionCreate(savedInstanceState, R.layout.bookmarks);

        onNewIntent(getIntent());

        ListView view = (ListView) findMyViewById(R.id.bookmarks);
        view.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                Bookmark bookmark = (Bookmark) parent.getItemAtPosition(position);
                Intent result = new Intent();
                result.putExtra(OPEN_PAGE, bookmark.page);
                System.out.println("bookmark id = " + bookmark.id + " page = " + bookmark.page);
                setResult(Activity.RESULT_OK, result);
                finish();
            }
        });
    }

    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        bookId = intent.getLongExtra(BOOK_ID, -1);
        updateView(bookId);
    }

    private void updateView(final long bookId) {
        BookmarkAccessor accessor = getOrionContext().getBookmarkAccessor();
        List bookmarks = accessor.selectBookmarks(bookId);
        final ListView view = (ListView) findMyViewById(R.id.bookmarks);
        view.setAdapter(new ArrayAdapter(this, R.layout.bookmark_entry, R.id.bookmark_entry, bookmarks) {
            @Override
            public View getView(final int position, View convertView, ViewGroup parent) {
                convertView = super.getView(position, convertView, parent);

                Bookmark item = (Bookmark) getItem(position);
                TextView page = (TextView) convertView.findViewById(R.id.bookmark_entry_page);
                page.setText("" + (item.page == -1 ? "*" : item.page + 1));

                ImageView edit = (ImageView) convertView.findViewById(R.id.bookmark_edit_entry);
                //if (edit != null)
                edit.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if (position != 0) {
                            final Bookmark item = (Bookmark) getItem(position);

                            AlertDialog.Builder builder = createThemedAlertBuilder().setIcon(R.drawable.edit_item);

                            builder.setTitle("Edit Bookmark");
                            final EditText editText = new EditText(OrionBookmarkActivity.this);
                            editText.setText(item.text);
                            builder.setView(editText);

                            builder.setPositiveButton("Save", new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog, int which) {
                                    if ("".equals(editText.getText())) {
                                        OrionBookmarkActivity.this.showAlert("Warning", "Coudn't save empty bookmark");
                                    } else {
                                        getOrionContext().getBookmarkAccessor().insertOrUpdateBookmark(bookId, item.page, editText.getText().toString());
                                        item.text = editText.getText().toString();
                                        notifyDataSetChanged();
                                    }
                                  dialog.dismiss();
                                }
                            });

                            builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog, int which) {
                                    dialog.dismiss();
                                }
                            });

                            builder.setNeutralButton("Delete", new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog, int which) {
                                    getOrionContext().getBookmarkAccessor().deleteBookmark(item.id);
                                    updateView(bookId);
                                }
                            });

                            builder.create().show();
                        }
                    }
                });
                return convertView;
            }
        });
    }

    @Override
    public boolean supportDevice() {
        return false;
    }

    public boolean onCreateOptionsMenu(Menu menu) {
        boolean result = super.onCreateOptionsMenu(menu);
        if (result) {
            getMenuInflater().inflate(R.menu.bookmarks_menu, menu);
        }
        return result;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        boolean showEmptyResult = false;

        switch (item.getItemId()) {

            case R.id.close_bookmarks_menu_item:
                finish();
                return true;

            case R.id.export_bookmarks_menu_item:
                if (bookId == -1) {
                    showEmptyResult = true;
                }

            case R.id.export_all_bookmarks_menu_item:
                String file = getOrionContext().getTempOptions().openedFile;
                if (file == null) {
                    showEmptyResult = true;
                }

                if (!showEmptyResult) {
                    long bookId = item.getItemId() == R.id.export_all_bookmarks_menu_item ? -1 : this.bookId;
                    file = file + "." + (bookId == -1 ? "all_" : "") +  "bookmarks.xml";
                    log("Bookmarks output file: " + file);

                    BookmarkExporter exporter = new BookmarkExporter(getOrionContext().getBookmarkAccessor(), file);
                    try {
                        showEmptyResult = !exporter.export(bookId);
                    } catch (IOException e) {
                        showError(e);
                        return true;
                    }
                }

                if (showEmptyResult) {
                    showLongMessage("There is nothing to export!");
                } else {
                    showLongMessage("Bookmarks exported to " + file);
                }
                return true;

            case R.id.import_current_bookmarks_menu_item:
                Intent intent = new Intent(this, OrionFileSelectorActivity.class);
                startActivityForResult(intent, IMPORT_CURRRENT);
                return true;

            case R.id.import_all_bookmarks_menu_item:
                intent = new Intent(this, OrionFileSelectorActivity.class);
                startActivityForResult(intent, IMPORT_ALL);
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            final String fileName = data.getStringExtra(OrionFileSelectorActivity.RESULT_FILE_NAME);
            if (fileName == null || "".equals(fileName)) {
                showWarning("File name is empty");
                return;
            } else {
                log("To import " + fileName);
                List<BookNameAndSize> books = listBooks(fileName);
                if (books == null) {
                    return;
                }

                if (books.isEmpty()) {
                    showWarning("There is no any bookmarks");
                }

                final boolean importCurrent = requestCode == IMPORT_CURRRENT;

                Collections.sort(books);

                View group = getLayoutInflater().inflate(R.layout.bookmark_book_list, null);
                final ListView tree = (ListView) group.findViewById(R.id.book_list);

                AlertDialog.Builder builder = createThemedAlertBuilder();
                builder.setTitle("Select source book").setCancelable(true).setView(group);
                builder.setPositiveButton("Import", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                        BookNameAndSize toBook = new BookNameAndSize(getOrionContext().getCurrentBookParameters().simpleFileName, getOrionContext().getCurrentBookParameters().fileSize);
                        doImport(fileName, getCheckedItems(tree), importCurrent ? toBook : null);
                    }
                });
                builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                });


                final AlertDialog dialog = builder.create();


                tree.setChoiceMode(!importCurrent ? ListView.CHOICE_MODE_MULTIPLE : ListView.CHOICE_MODE_SINGLE);
                tree.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                        dialog.getButton(Dialog.BUTTON_POSITIVE).setEnabled(tree.getCheckedItemPositions().size() != 0);
                    }
                });

                tree.setAdapter(new ArrayAdapter<BookNameAndSize>(this, importCurrent ? R.layout.select_book_item : R.layout.select_book_item_multi, R.id.title, books) {

                    private boolean positiveDisabled;

                    @Override
                    public View getView(int position, View convertView, ViewGroup parent) {
                        convertView = super.getView(position, convertView, parent);

                        BookNameAndSize book = getItem(position);
                        TextView view = (TextView) convertView.findViewById(R.id.page);
                        view.setText(book.buityfySize());

                        view = (TextView) convertView.findViewById(R.id.title);
                        view.setText(book.getName());

                        if (!positiveDisabled) {
                            //android bug
                            positiveDisabled = true;
                            if (importCurrent) {
                                dialog.getButton(Dialog.BUTTON_POSITIVE).setEnabled(false);
                            }
                        }

                        return convertView;
                    }
                });
                tree.setItemsCanFocus(false);

                if (!importCurrent) {
                    for (int i = 0; i < tree.getAdapter().getCount(); i++) {
                          tree.setItemChecked(i, true);
                    }
                }

                dialog.show();

            }
        }
    }

    private void doImport(String fileName, Set<BookNameAndSize> books, BookNameAndSize toBook){
        log("Import bookmarks " + books.size());

        BookmarkImporter importer = new BookmarkImporter(getOrionContext().getBookmarkAccessor(), fileName, books, toBook);
        try {
            importer.doImport();
            updateView(getOrionContext().getBookmarkAccessor().selectBookId(getOrionContext().getCurrentBookParameters().simpleFileName, getOrionContext().getCurrentBookParameters().fileSize));
            showFastMessage("Imported successfully");
        } catch (OrionException e) {
            showAlert("Error", e.getMessage());
        }

    }


    public List<BookNameAndSize> listBooks(String fileName) {
        ArrayList<BookNameAndSize> bookNames = new ArrayList<BookNameAndSize>();
        InputStreamReader reader = null;
        try {
            reader = new InputStreamReader(new FileInputStream(new File(fileName)));
            XmlPullParserFactory factory = XmlPullParserFactory.newInstance();

            XmlPullParser xpp = factory.newPullParser();
            xpp.setInput(reader);

            int eventType = xpp.getEventType();
            boolean wasBookmarks = false;

            while (eventType != XmlPullParser.END_DOCUMENT) {
                if (eventType == XmlPullParser.START_TAG) {
                    String tagName = xpp.getName();
                    if ("bookmarks".equals(tagName)) {
                        wasBookmarks = true;
                    } else if (wasBookmarks) {
                        if ("book".equals(tagName)) {
                            String name = xpp.getAttributeValue(NAMESPACE, "fileName");
                            long size = Long.valueOf(xpp.getAttributeValue(NAMESPACE, "fileSize"));
                            BookNameAndSize book = new BookNameAndSize(name, size);
                            bookNames.add(book);
                        }
                    }
                }
                eventType = xpp.next();
            }
            return bookNames;
        } catch (FileNotFoundException e) {
            showError("Couldn't open file", e);
        } catch (XmlPullParserException e) {
            showError("Couldn't parse book parameters", e);
        } catch (IOException e) {
            showError("Couldn't parse book parameters", e);
        } finally {
            if (reader != null) {
                try {
                    reader.close();
                } catch (IOException e) {
                    log(e);
                }
            }
        }
        return null;
    }

    private Set<BookNameAndSize> getCheckedItems(ListView view) {
        HashSet<BookNameAndSize> result = new HashSet();
        SparseBooleanArray checked = view.getCheckedItemPositions();
        for (int i = 0; i < checked.size(); i++) {
            if (checked.valueAt(i)) {
                result.add((BookNameAndSize)view.getAdapter().getItem(checked.keyAt(i)));
            }
        }
        return  result;
    }

}

File: orion-viewer/src/main/java/universe/constellation/orion/viewer/OrionBookmarkActivity.kt
/*
 * Orion Viewer - pdf, djvu, xps and cbz file viewer for android devices
 *
 * Copyright (C) 2011-2013  Michael Bogdanov & Co
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package universe.constellation.orion.viewer

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Dialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.net.toUri
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import org.xmlpull.v1.XmlPullParserFactory
import universe.constellation.orion.viewer.Permissions.checkWritePermission
import universe.constellation.orion.viewer.bookmarks.BookNameAndSize
import universe.constellation.orion.viewer.bookmarks.Bookmark
import universe.constellation.orion.viewer.bookmarks.ALL_BOOKMARKS_SUFFIX
import universe.constellation.orion.viewer.bookmarks.BOOKMARKS_SUFFIX
import universe.constellation.orion.viewer.bookmarks.BookmarkExporter
import universe.constellation.orion.viewer.bookmarks.BookmarkImporter
import universe.constellation.orion.viewer.bookmarks.bookmarksExportFile
import java.io.*
import java.util.*

class OrionBookmarkActivity : OrionBaseActivity() {

    private var bookId: Long = 0

    @SuppressLint("MissingSuperCall")
    public override fun onCreate(savedInstanceState: Bundle?) {
        onOrionCreate(savedInstanceState, R.layout.bookmarks,
            displayHomeAsUpEnabled = true
        )

        onNewIntent(intent)

        val view = findViewById<ListView>(R.id.bookmarks)
        view.onItemClickListener = AdapterView.OnItemClickListener { parent, _, position, _ ->
            val bookmark = parent.getItemAtPosition(position) as Bookmark
            val result = Intent()
            result.putExtra(OPEN_PAGE, bookmark.page)
            log("bookmark id = " + bookmark.id + " page = " + bookmark.page)
            setResult(Activity.RESULT_OK, result)
            finish()
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        bookId = intent.getLongExtra(BOOK_ID, -1)
        updateView(bookId)
    }

    private fun updateView(bookId: Long) {
        val accessor = orionApplication.getBookmarkAccessor()
        val bookmarks = accessor.selectBookmarks(bookId)
        val view = findViewById<ListView>(R.id.bookmarks)
        view.adapter = object : ArrayAdapter<Bookmark>(this, R.layout.bookmark_entry, R.id.bookmark_entry, bookmarks) {
            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                val convertView = super.getView(position, convertView, parent)

                val bookmark = getItem(position)
                val page = convertView.findViewById<View>(R.id.bookmark_entry_page) as TextView
                page.text = "${if (bookmark!!.page == -1) "*" else bookmark.page + 1}"

                val edit = convertView.findViewById<View>(R.id.bookmark_edit_entry) as ImageView
                //if (edit != null)
                edit.setOnClickListener {
                    if (position != 0) {
                        val item = getItem(position)!!

                        val builder = createThemedAlertBuilder().setIcon(R.drawable.edit_item)

                        builder.setTitle("Edit Bookmark")
                        val editText = EditText(this@OrionBookmarkActivity)
                        editText.setText(item.text)
                        builder.setView(editText)

                        builder.setPositiveButton("Save") { dialog, which ->
                            if (editText.text.isEmpty()) {
                                this@OrionBookmarkActivity.showAlert("Warning", "Coudn't save empty bookmark")
                            } else {
                                orionApplication.getBookmarkAccessor().insertOrUpdateBookmark(bookId, item.page, editText.text.toString())
                                item.text = editText.text.toString()
                                notifyDataSetChanged()
                            }
                            dialog.dismiss()
                        }

                        builder.setNegativeButton("Cancel") { dialog, _ -> dialog.dismiss() }

                        builder.setNeutralButton("Delete") { _, _ ->
                            orionApplication.getBookmarkAccessor().deleteBookmark(item.id.toLong())
                            updateView(bookId)
                        }

                        builder.create().show()
                    }
                }
                return convertView
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val result = super.onCreateOptionsMenu(menu)
        if (result) {
            menuInflater.inflate(R.menu.bookmarks_menu, menu)
        }
        return result
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {

            R.id.export_bookmarks_menu_item -> {
                //should be granted automatically
                checkWritePermission(this)
                if (bookId == -1L) {
                    showLongMessage("There is nothing to export!")
                } else {
                    exportBookmarks(bookId, BOOKMARKS_SUFFIX)
                }
                return true
            }

            R.id.export_all_bookmarks_menu_item -> {
                //should be granted automatically
                checkWritePermission(this)
                exportBookmarks(EXPORT_ALL_BOOKS, ALL_BOOKMARKS_SUFFIX)
                return true
            }

            R.id.import_current_bookmarks_menu_item -> {
                val intent = Intent(this, OrionFileSelectorActivity::class.java)
                startActivityForResult(intent, IMPORT_CURRRENT)
                return true
            }

            R.id.import_all_bookmarks_menu_item -> {
                intent = Intent(this, OrionFileSelectorActivity::class.java)
                startActivityForResult(intent, IMPORT_ALL)
                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK) {
            val fileUri = data?.getStringExtra(OrionFileSelectorActivity.RESULT_FILE_URI)
                ?.takeIf { it.isNotEmpty() }?.toUri()
            if (fileUri == null) {
                showWarning("File name is empty")
                return
            } else {
                log("To import $fileUri")
                val books = listBooks(fileUri) ?: return

                if (books.isEmpty()) {
                    showWarning("There is no any bookmarks")
                    return
                }

                val importCurrent = requestCode == IMPORT_CURRRENT

                books.sort()

                val group = layoutInflater.inflate(R.layout.bookmark_book_list, null)
                val tree = group.findViewById<View>(R.id.book_list) as ListView

                val builder = createThemedAlertBuilder()
                builder.setTitle("Select source book").setCancelable(true).setView(group)
                builder.setPositiveButton("Import") { dialog, which ->
                    dialog.dismiss()
                    val currentBookParameters = orionApplication.currentBookParameters
                    if (importCurrent && currentBookParameters == null) {
                        showWarning("There is no opened book to import into")
                        return@setPositiveButton
                    }
                    val toBook = currentBookParameters?.let {
                        BookNameAndSize(it.simpleFileName, it.fileSize)
                    }
                    doImport(fileUri, getCheckedItems(tree), if (importCurrent) toBook else null)
                }
                builder.setNegativeButton("Cancel") { dialog, _ -> dialog.dismiss() }


                val dialog = builder.create()


                tree.choiceMode = if (!importCurrent) ListView.CHOICE_MODE_MULTIPLE else ListView.CHOICE_MODE_SINGLE
                tree.onItemClickListener = AdapterView.OnItemClickListener { _, _, _, _ -> dialog.getButton(Dialog.BUTTON_POSITIVE).isEnabled = tree.checkedItemPositions.size() != 0 }

                tree.adapter = object : ArrayAdapter<BookNameAndSize>(this, if (importCurrent) R.layout.select_book_item else R.layout.select_book_item_multi, R.id.title, books) {

                    private var positiveDisabled: Boolean = false

                    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                        val convertView = super.getView(position, convertView, parent)

                        val book = getItem(position)
                        var view = convertView.findViewById<View>(R.id.page) as TextView
                        view.text = book!!.beautifySize()

                        view = convertView.findViewById<View>(R.id.title) as TextView
                        view.text = book.name

                        if (!positiveDisabled) {
                            //android bug
                            positiveDisabled = true
                            if (importCurrent) {
                                dialog.getButton(Dialog.BUTTON_POSITIVE).isEnabled = false
                            }
                        }

                        return convertView
                    }
                }
                tree.itemsCanFocus = false

                if (!importCurrent) {
                    for (i in 0 until tree.adapter.count) {
                        tree.setItemChecked(i, true)
                    }
                }

                dialog.show()

            }
        }
    }

    private fun exportBookmarks(bookId: Long, suffix: String) {
        val target = bookmarksExportFile(orionApplication.currentBookParameters?.openingFileName, suffix)
        log("Bookmarks output file: $target")

        val exporter = BookmarkExporter(orionApplication.getBookmarkAccessor(), target.absolutePath)
        val exported = try {
            exporter.export(bookId)
        } catch (e: IOException) {
            showAndLogError(this, e)
            return
        }

        if (exported) {
            showLongMessage("Bookmarks exported to $target")
        } else {
            showLongMessage("There is nothing to export!")
        }
    }

    private fun doImport(fileUri: Uri, books: Set<BookNameAndSize>, toBook: BookNameAndSize?) {
        log("Import bookmarks " + books.size)

        val input = openImportStream(fileUri) ?: return
        val importer = BookmarkImporter(orionApplication.getBookmarkAccessor(), input, books, toBook)
        try {
            importer.doImport()
            val currentBookParameters = orionApplication.currentBookParameters
            val bookId = currentBookParameters?.let {
                orionApplication.getBookmarkAccessor().selectBookId(it.simpleFileName, it.fileSize)
            } ?: -1L
            updateView(bookId)
            showFastMessage("Imported successfully")
        } catch (e: OrionException) {
            showAlert("Error", e.message!!)
        }

    }

    private fun openImportStream(fileUri: Uri): InputStream? {
        return try {
            contentResolver.openInputStream(fileUri)
                ?: throw FileNotFoundException("No content for $fileUri")
        } catch (e: Exception) {
            showAndLogError(this, "Couldn't open file", e)
            null
        }
    }

    private fun listBooks(fileUri: Uri): MutableList<BookNameAndSize>? {
        val bookNames = ArrayList<BookNameAndSize>()
        var reader: InputStreamReader? = null
        try {
            reader = InputStreamReader(openImportStream(fileUri) ?: return null)
            val factory = XmlPullParserFactory.newInstance()

            val xpp = factory.newPullParser()
            xpp.setInput(reader)

            var eventType = xpp.eventType
            var wasBookmarks = false

            while (eventType != XmlPullParser.END_DOCUMENT) {
                if (eventType == XmlPullParser.START_TAG) {
                    val tagName = xpp.name
                    if ("bookmarks" == tagName) {
                        wasBookmarks = true
                    } else if (wasBookmarks) {
                        if ("book" == tagName) {
                            val name = xpp.getAttributeValue(NAMESPACE, "fileName")
                            val size = java.lang.Long.valueOf(xpp.getAttributeValue(NAMESPACE, "fileSize"))
                            val book = BookNameAndSize(name, size)
                            bookNames.add(book)
                        }
                    }
                }
                eventType = xpp.next()
            }
            return bookNames
        } catch (e: FileNotFoundException) {
            showAndLogError(this, "Couldn't open file", e)
        } catch (e: XmlPullParserException) {
            showAndLogError(this, "Couldn't parse book parameters", e)
        } catch (e: IOException) {
            showAndLogError(this, "Couldn't parse book parameters", e)
        } finally {
            if (reader != null) {
                try {
                    reader.close()
                } catch (e: IOException) {
                    log(e)
                }

            }
        }
        return null
    }

    private fun getCheckedItems(view: ListView): Set<BookNameAndSize> {
        val result = HashSet<BookNameAndSize>()
        val checked = view.checkedItemPositions
        for (i in 0 until checked.size()) {
            if (checked.valueAt(i)) {
                result.add(view.adapter.getItem(checked.keyAt(i)) as BookNameAndSize)
            }
        }
        return result
    }

    companion object {

        const val OPEN_PAGE = "open_page"

        const val OPENED_FILE = "opened_file"

        const val BOOK_ID = "book_id"

        const val NAMESPACE = ""

        const val IMPORT_CURRRENT = 1

        const val IMPORT_ALL = 2

        const val EXPORT_ALL_BOOKS = -1L
    }

}
